designs/0005-static-matrices.md

```diff
@@ -0,0 +1,252 @@
+- Feature Name: static_matrices
+- Start Date: April 30th, 2020
+- RFC PR: [Example Branch](https://github.com/stan-dev/math/compare/feature/var-template)
+- Stan Issue: [#1805](https://github.com/stan-dev/math/issues/1805)
+
+# Summary
+[summary]: #summary
+
+This proposes two things:
+
+1. A generalization of the C++ reverse mode autodiff type in Stan Math, from `stan::math::var`
+to `stan::math::var_value<T>`, where `T` is the underlying storage type of the autodiff
+variable. For `T == double`, this type is equivalent to the current Stan Math `var` type.
+The usefulness of the generalized reverse mode autodiff type comes from passing more
+general data structures as the template type, in particular the Eigen Matrix types.
+
+2. A design for exposing these autodiff variables at the Stan language level
+
+An autodiff type that allows customization of the underlying storage object makes it possible
+for matrices, instead of being data structures of autodiff types, for the autodiff type
+itself to represent a matrix. Operations on these new types can be more efficient than the old
+ones, and this design remains fully backwards compatible with the curren Stan autodiff types.
+
+For the rest of this proposal, "autodiff" means "reverse mode autodiff," and all C++ functions
+are assumed to reside in the `stan::math` namespace. The necessary namespace will be
+included if this is not true.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, an `N x M` matrix in Stan is represented as an Eigen matrix holding
+a pointer to an underlying array of autodiff objects, each of those holding
+a pointer to the underlying autodiff object implementation.
+
+The basic Eigen type used by Stan is `Eigen::Matrix<var, R, C>`. For brevity,
+this will be referred to as a `matvar` type (a matrix containing var variables).
+There are no major differences between a matrix of vars or a vector of vars or
+a row vector of vars.
+
+This proposal will enable an `N x M` Stan matrix to be defined in terms of one
+autodiff object with two separate `N x M` containers, one for its values and
+one for its adjoints.
```

Comment:
I think this would clean up a bit with some pseudocode examples. I think you can take (1) and (2) that you have written below and include them in the example for matrix multiplication

-------------------------

Currently, Stan represents matrices as an `Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>`. There are no major differences between a matrix of vars or a vector of vars or a row vector of vars. The internal representation of an Eigen matrix is a pointer to an array.

```cpp
// pseudo class representing the storage of a dense matrix
template <typename T>
class DenseStorage {
  T* data_;  
};
```

Stan math's reverse mode autodiff (`var`) is a smart pointer type that itself points to an underlying object (`vari`) that actually holds the values and adjoints that Stan math uses when calculating gradients.

```cpp
// pseudo class holding the values and adjoints used in reverse mode AD
class vari {
  double val_;
  double adj_;
};

// pseudoclass that represents that stores the pointer to the underlying vari
class var {
  vari* vi_;
}
```

This means that for Eigen matrices of type `Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>` it's internal representation is a pointer to an array of pointers.

```
// pseudo class that shows equivalently what the compiler sees
class DenseStorage {
  vari** data_;  
}
```

This becomes problematic with both pointer chasing and memory access. Typically when Stan math performs operations over matrices it will want all of the values of a `vari` for one part of an operation and the all the adjoints of another `vari` in another part of an operation. This can be wasteful as to fetch just the values or adjoints we need to go through the pointer of pointers and load both the values and adjoints into cache. And unless the value and adjoints are used together in the same operation then the CPU is also seeing non-contiguous data.

This proposal will enable an `N x M` Stan matrix to be defined in terms of one
autodiff object with two separate `N x M` containers, one for its values and
one for its adjoints. This is represented as a `var_value<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>>`.

```cpp
// pseudo class for matrix vari
template <typename T>
class vari_value {
  Eigen::Map<T>* val_;
  Eigen::Map<T>* adj_;
}
// pseudo code representing var matrix
template <typename T>
class var_value {
  vari_value<T>* vi_;
};
```

This storage scheme allows for less pointer chasing while providing contiguous memory access to values and adjoints. For instance, when performing the reverse mode pass for matrix autodiff the calculation for propagating the adjoints of each input requires accessing the values of the other input

```cpp
A.adj() += result.adj() * B.val().transpose();
B.adj() += A.val().transpose() * result.adj();
```

With an Eigen matrix of var types this means that each multiplication in the above requires the program to fetch both the values and adjoints from the pointer to the array of pointers for both `A` and `B` when calculating either adjoint. Furthermore this makes those values non-continguous in the operation and so the compiler will not be able to use advanced instruction sets such as SIMD multiplication.

However, switching to a `var_value` with the values and adjoints separated into two arrays means that both adjoint calculations will only be fetching the values or adjoints they need in the operation. And those values and adjoints will be contiguous so that the compiler has a much better opportunity to utilize advanced instruction sets when performing the calculation.

Reply:
Thanks, I agree the C++ is useful for showing what the types are. I reordered some stuff and dropped the DenseStorage things.

Reply:
Why did you drop the densestorage? I think it's one of the bigger points that the vari** impl causes a bunch of pointer chasing. I prefer just having that laid out simply because idt it's obvious at first

Reply:
Fair, I'll add something back.